src/HttpServer.php

```diff
@@ -125,18 +125,18 @@ public function handleBody(RequestInterface $request, ConnectionInterface $conne
             return;
         }
 
-        $listener = function ($data) use ($request, &$bodyBuffer, $connection, $that, &$listener) {
-            $bodyBuffer .= $data;
-            $contentLengthArray = $request->getHeader('Content-Length');
+        $stream = new LengthLimitedStream($connection, (int)$request->getHeaderLine('Content-Length'));
 
-            if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                $request = $request->withBody(RingCentral\Psr7\stream_for($bodyBuffer));
-                $connection->removeListener('data', $listener);
-                $that->handleRequest($connection, $request);
-            }
-        };
+        $bodyBuffer = '';
 
-        $connection->on('data', $listener);
+        $stream->on('data', function ($data) use (&$bodyBuffer) {
+            $bodyBuffer .= $data;
+        });
+
+        $stream->on('end', function () use (&$bodyBuffer, &$listener, $connection, $that, $request) {
+            $request = $request->withBody(RingCentral\Psr7\stream_for($bodyBuffer));
+            $that->handleRequest($connection, $request);
```

Comment:
Functionally LGTM :+1: However, the exact same code also appears above for the chunked decoder? (DRY)